include/xgboost/data.h

```diff
@@ -367,11 +378,10 @@ class DMatrix {
   /*! \brief meta information of the dataset */
   virtual const MetaInfo& Info() const = 0;
   /**
-   * \brief Gets row batches. Use range based for loop over BatchSet to access individual batches.
+   * \brief Gets batches. Use range based for loop over BatchSet to access individual batches.
    */
-  virtual BatchSet GetRowBatches() = 0;
-  virtual BatchSet GetSortedColumnBatches() = 0;
-  virtual BatchSet GetColumnBatches() = 0;
+  template<typename T>
+  BatchSet<T> GetBatches(PageType page_type);
```

Comment:
natural type deduction for function templates will not occur if the return type is based on the template type, thus forcing the call site to explicitly provide the type name when this method is invoked. it looks more natural to provide a type when an iterator is retrieved with an appropriate type than this version.

Reply:
I guess another possibility is to pass some kind of prototype instance for type inference
```cpp
template<typename T>
BatchSet<T> GetBatches(T prototype);
...
// caller
for (auto& batch: dmat->GetBatches(SparsePage::Instance) {
...
}
```
but that seems a little baroque.